internal/credential/vault/repository_credential_store.go

```diff
@@ -718,50 +710,78 @@ func (r *Repository) UpdateCredentialStore(ctx context.Context, cs *CredentialSt
 				}
 				if rows > 1 {
-					return errors.New(errors.MultipleRecords, op, "more than 1 token would have been created")
+					return errors.New(errors.MultipleRecords, op, "more than 1 client certificate would have been deleted")
 				}
-				msgs = append(msgs, returnedToken.oplogMessage(db.CreateOp))
-
-				returnedCredentialStore.inputToken = nil
-				returnedToken.Token.Token = nil
-				returnedToken.Token.CtToken = nil
-				returnedCredentialStore.outputToken = returnedToken
-			}
-
-			switch {
-			case deleteClientCert:
-				returnedClientCert = clientCert.clone()
-				query, values := returnedClientCert.deleteQuery()
+				msgs = append(msgs, deleteCert.oplogMessage(db.DeleteOp))
+			case len(certDbMask) > 0:
+				if updatedStore.clientCert == nil {
+					return errors.New(errors.InvalidParameter, op, "updated cert")
+				}
+				updatedStore.clientCert.StoreId = ps.StoreId
+				updatedStore.clientCert.encrypt(ctx, databaseWrapper)
+				query, values := updatedStore.clientCert.insertQuery()
 				rows, err := w.Exec(ctx, query, values)
 				if err != nil {
-					return errors.Wrap(err, op, errors.WithMsg("unable to delete client certificate"))
+					return errors.Wrap(err, op, errors.WithMsg("unable to upsert client certificate"))
 				}
 				if rows > 1 {
-					return errors.New(errors.MultipleRecords, op, "more than 1 client certificate would have been deleted")
+					return errors.New(errors.MultipleRecords, op, "more than 1 client certificate would have been upserted")
+				}
+			}
+
+			var token *Token
+			if updateToken {
+				client, err := updatedStore.client()
+				if err != nil {
+					return errors.Wrap(err, op, errors.WithMsg("unable to get client for updated store"))
 				}
-				msgs = append(msgs, returnedClientCert.oplogMessage(db.DeleteOp))
-			case updateClientCert:
-				returnedClientCert = clientCert.clone()
-				query, values := returnedClientCert.insertQuery()
+				tokenLookup, err := client.lookupToken()
+				if err != nil {
+					return errors.Wrap(err, op, errors.WithMsg("cannot lookup token for updated store"))
+				}
+				if err := validateTokenLookup(op, tokenLookup); err != nil {
+					return err
+				}
+				renewedToken, err := client.renewToken()
+				if err != nil {
+					return errors.Wrap(err, op, errors.WithMsg("unable to renew vault token"))
+				}
+				tokenExpires, err := renewedToken.TokenTTL()
+				if err != nil {
+					return errors.Wrap(err, op, errors.WithMsg("unable to get vault token expiration"))
+				}
+				accessor, err := renewedToken.TokenAccessor()
+				if err != nil {
+					return errors.Wrap(err, op, errors.WithMsg("unable to get vault token accessor"))
+				}
+				token, err = newToken(cs.GetPublicId(), cs.inputToken, []byte(accessor), tokenExpires)
+				if err != nil {
+					return err
+				}
+				// encrypt token
+				if err := token.encrypt(ctx, databaseWrapper); err != nil {
+					return errors.Wrap(err, op)
+				}
+				query, values := token.insertQuery()
```

Comment:
Why move all of this within the scope of the transaction? In general, we should try to keep the time of a database transaction as short as possible?

Reply:
Moved a bunch of stuff outside of the transaction.

Reply:
ty